File: data_structure/binaryTree/inorder/inoder_traversal_recursion.py

# 二叉树的中序遍历

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# 迭代
def inorderTraversal(root: TreeNode) ->[int]:
    result = []
    queue = []
    while root or queue :
        while root :
            queue.append(root)
            root = root.left
        node = queue.pop()
        result.append(node.val)
        root = node.right
    return result



if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(2)

    root.left.left = TreeNode(3)
    root.left.right = TreeNode(4)

    root.right.left = TreeNode(4)
    root.right.right = TreeNode(3)
    result = inorderTraversal(root)
    print(result)

File: data_structure/binaryTree/inorder/inoder_traversal_recursion.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {

	root := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 2}}
	root.Left.Left = &TreeNode{Val: 3}
	root.Left.Right = &TreeNode{Val: 4}
	root.Right.Left = &TreeNode{Val: 4}
	root.Right.Right = &TreeNode{Val: 3}
	result := inorderTraversal(root)
	fmt.Println(result)

}

// 递归 中序遍历二叉树
func inorderTraversal(root *TreeNode) []int {
	result := []int{}
	if root == nil {
		return result
	}
	left := inorderTraversal(root.Left)
	if len(left) != 0 {
		result = append(result, left...)
	}
	result = append(result, root.Val)
	right := inorderTraversal(root.Right)
	if len(right) != 0 {
		result = append(result, right...)
	}
	return result
}
